File: melanoma/src/infrastructure/vertex_lock.py
"""One bulk Vertex AI workload per project on this machine.

The pipelines here and the web app's golden evals (``web/src/lib/agent/evals``)
draw on the same project's shared-pool standing. Two of them at once is how a
run gets refused for reasons that have nothing to do with it, so each takes
this lock for its whole process and a second one refuses to start.

The lock is a file holding the owner's pid. A file whose pid is no longer
running was left by a crash and is taken over. The web side reads and writes
the same file (``vertex-lock.ts``), so keep the format in step with it.
"""

from __future__ import annotations

import atexit
import json
import os
from datetime import datetime, timezone
from pathlib import Path

LOCK_DIR = Path.home() / ".bionocular"


class VertexWorkloadBusyError(RuntimeError):
    """Another bulk workload holds this project's lock."""


def _alive(pid: int) -> bool:
    try:
        os.kill(pid, 0)
    except ProcessLookupError:
        return False
    except PermissionError:
        return True
    return True


def hold_vertex_lock(project: str, workload: str, lock_dir: Path = LOCK_DIR) -> Path:
    """Take the project's lock until this process exits, or raise if it is held."""
    lock_dir.mkdir(parents=True, exist_ok=True)
    path = lock_dir / f"vertex-{project}.lock"
    owner = json.dumps(
        {
            "pid": os.getpid(),
            "workload": workload,
            "startedAt": datetime.now(timezone.utc).isoformat(),
        }
    )
    for _ in range(2):
        try:
            fd = os.open(path, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o644)
        except FileExistsError:
            try:
                holder = json.loads(path.read_text())
            except (OSError, ValueError):
                holder = {}
            pid = holder.get("pid")
            if pid == os.getpid():
                return path
            if isinstance(pid, int) and _alive(pid):
                raise VertexWorkloadBusyError(
                    f"{holder.get('workload', 'another workload')} (pid {pid}, since "
                    f"{holder.get('startedAt', '?')}) is already using Vertex project "
                    f"{project}. Wait for it to finish; if that process is gone, "
                    f"delete {path}."
                ) from None
            path.unlink(missing_ok=True)
            continue
        with os.fdopen(fd, "w") as f:
            f.write(owner)
        atexit.register(_release, path)
        return path
    raise VertexWorkloadBusyError(
        f"Could not take {path}; another process raced for it."
    )


def _release(path: Path) -> None:
    try:
        if json.loads(path.read_text()).get("pid") == os.getpid():
            path.unlink()
    except (OSError, ValueError):
        pass

File: web/src/lib/agent/evals/vertex-lock.ts
/**
 * One bulk Vertex AI workload per project on this machine.
 *
 * The golden evals and the melanoma pipelines draw on the same project's
 * shared-pool standing, so two at once get refused for each other's load.
 * Each takes this lock for its run and a second one refuses to start. The
 * file and its format are shared with `melanoma/src/infrastructure/vertex_lock.py`;
 * keep the two in step.
 */

import { mkdirSync, openSync, readFileSync, rmSync, writeSync, closeSync } from 'node:fs';
import { homedir } from 'node:os';
import path from 'node:path';

export const LOCK_DIR = path.join(homedir(), '.bionocular');

interface Holder {
  pid?: number;
  workload?: string;
  startedAt?: string;
}

function alive(pid: number): boolean {
  try {
    process.kill(pid, 0);
    return true;
  } catch (error) {
    return (error as NodeJS.ErrnoException).code === 'EPERM';
  }
}

function readHolder(file: string): Holder {
  try {
    return JSON.parse(readFileSync(file, 'utf8')) as Holder;
  } catch {
    return {};
  }
}

/** Take the project's lock, or throw naming whoever holds it. Returns the release. */
export function holdVertexLock(project: string, workload: string, dir = LOCK_DIR): () => void {
  mkdirSync(dir, { recursive: true });
  const file = path.join(dir, `vertex-${project}.lock`);
  const owner = JSON.stringify({ pid: process.pid, workload, startedAt: new Date().toISOString() });
  const release = () => {
    if (readHolder(file).pid === process.pid) rmSync(file, { force: true });
  };

  for (let attempt = 0; attempt < 2; attempt++) {
    let fd: number;
    try {
      fd = openSync(file, 'wx');
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code !== 'EEXIST') throw error;
      const holder = readHolder(file);
      if (holder.pid === process.pid) return release;
      if (typeof holder.pid === 'number' && alive(holder.pid)) {
        throw new Error(
          `${holder.workload ?? 'another workload'} (pid ${holder.pid}, since ${holder.startedAt ?? '?'}) ` +
            `is already using Vertex project ${project}. Wait for it to finish; if that process is gone, delete ${file}.`,
        );
      }
      rmSync(file, { force: true });
      continue;
    }
    writeSync(fd, owner);
    closeSync(fd);
    return release;
  }
  throw new Error(`Could not take ${file}; another process raced for it.`);
}
